### PR TITLE
feat: Agentic Improvements

### PR DESCRIPTION
**Pull request by ginccc**
## Summary

This PR evolves EDDI from a configured pipeline to a **governed agent ecosystem**. Agents can now discover each other's capabilities, carry multimodal context, self-regulate behavior through counterweights, recover from failed operations via memory checkpoints, and prove their identity cryptographically.

All 9 features are **config-driven and off by default** — existing agents work unchanged.

**70 files changed** · **+7,118 / −280 lines** · 55 Java files · 23 test classes · 21 commits

---

### 🛡️ Behavioral Counterweights
Safety instructions injected into the system prompt at engine level. Three preset levels (`normal` → `cautious` → `strict`) with auto-downgrade in batch/scheduled runs. Customizable via Prompt Snippets without redeployment.

### 🎭 Identity Masking
Prevents agents from revealing their model name, provider, or architecture when probed. Configurable rules prepended to the system prompt.

### 📏 Tool Response Truncation (3 strategies)
Prevents context window blowout from verbose tool outputs:
- **`truncate`** — hard cut with note (default, backward-compatible)
- **`paginate`** — split into pages, LLM fetches more via `fetch_tool_response_page` built-in tool
- **`summarize`** — routes through a cheap model (inherits parent task's API key), falls back to truncate on failure. 6-guard fallback chain with cost ceiling, empty/oversized summary protection, and full log sanitization.

### 🔍 A2A Capability Registry
Runtime agent discovery via structured skills. Agents advertise capabilities with confidence levels; behavior rules query the registry with 4 selection strategies (`highest_confidence`, `round_robin`, `all`, `random`). Optional public discovery endpoint at `/.well-known/capabilities`.

### 💾 Memory Checkpoints & Rollback
Snapshot conversation state before risky tool executions, roll back cleanly on failure. Properties preserve full scope metadata (`longTerm`/`step`/`conversation`/`secret`) through rollback. Configurable retention (default: 10 checkpoints per conversation).

### 📎 Multimodal Attachments
Images, audio, and documents flow through the pipeline to vision-capable LLMs. Dual-backend storage (GridFS + PostgreSQL), magic-byte MIME validation, automatic `ImageContent`/`AudioContent` conversion.

### 🔐 Cryptographic Agent Identity
Ed25519 key generation, signing, and verification with vault-backed private keys. Key rotation support with versioned public keys and temporal validity windows. Security flags validated but intentionally not yet wired (HTTP 400 on premature activation).

### 📝 Memory Policy (Commit Flags)
Strict write discipline hides failed task data from subsequent LLM turns, injecting a concise error digest instead. Three modes: `digest`, `exclude_all`, `keep_all`.

### 🌍 Deployment Context Condition
Environment-aware behavior rules. Match on deployment environment (`production`/`staging`/`development`) and agent tags to auto-activate safety features per environment.

---

### Per-task (`langchain.json`)
```json
{
  "counterweight": { "enabled": true, "level": "cautious" },
  "identityMasking": { "enabled": true, "rules": ["..."] },
  "toolResponseLimits": {
    "defaultMaxChars": 5000,
    "truncationStrategy": "summarize",
    "summarizerModel": "gpt-4o-mini"
  }
}
```

### Per-agent (`AgentConfiguration`)
```json
{
  "capabilities": [{ "skill": "translation", "confidence": "high" }],
  "sessionManagement": { "autoSnapshot": { "enabled": true, "triggerOn": ["before_tool"] } },
  "memoryPolicy": { "strictWriteDiscipline": { "enabled": true, "onFailure": "digest" } }
}
```

---

## Testing

- **5,076 tests** in the full suite, **>90% coverage** on all new classes
- 23 new/modified test classes covering all features and fallback paths
- 21 pre-existing infrastructure test failures (Docker-dependent Mongo/Postgres) — unchanged from `main`

## Breaking Changes

**None.** All features default to off. Existing agent configurations, REST API contracts, and MongoDB document schemas are fully backward-compatible.

## Documentation

- [`docs/changelog.md`](docs/changelog.md) — detailed entries for every change with design rationale
- [`planning/manager-ui-handoff.md`](planning/manager-ui-handoff.md) — complete spec for Manager UI integration (JSON paths, field types, defaults, suggested widgets)
- [`planning/agentic-improvements-plan.md`](planning/agentic-improvements-plan.md) — original design plan with updated status

## Summary by CodeRabbit

* **New Features**
  * Conversation-scoped attachment upload/list/delete with MIME validation, size guarding, and attachment-backed multimodal message enhancement
  * Conversation checkpoints (create/rollback/prune) and session management controls
  * Per-task behavioral safety: identity masking and counterweight presets (prefix/suffix, levels)
  * Capability discovery endpoints and deterministic round‑robin selection; paginated/summarized tool responses with fetch-page tool
  * Agent message signing surfaced in group transcripts

* **Improvements**
  * PostgreSQL option for attachments and checkpoints; in‑memory paginated response cache

* **Documentation**
  * Architecture, changelog, langchain, planning and UI handoff docs updated

* **Tests**
  * Large suite of new and expanded unit tests across features and edge cases

[Review Change Stack](https://app.coderabbit.ai/change-stack/labsai/EDDI/pull/485)

### CODE REVIEW

File: src/main/java/ai/labs/eddi/engine/memory/DeepCopyUtil.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            Map<String, Object> copy = new LinkedHashMap<>(map.size());
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                copy.put(String.valueOf(entry.getKey()), deepCopyValue(entry.getValue()));
+            }
+            return copy;
+        }
+        if (value instanceof List<?> list) {
+            List<Object> copy = new ArrayList<>(list.size());
+            for (Object item : list) {
+                copy.add(deepCopyValue(item));
```
Comment (rolandpickl) [inline src/main/java/ai/labs/eddi/engine/memory/DeepCopyUtil.java]: I am always hesitant to have a full recursion without a maximum depth. Can we do such a thing?

**Review by rolandpickl — CHANGES_REQUESTED** [review]

File: src/main/java/ai/labs/eddi/engine/memory/DeepCopyUtil.java (hunk 1 above)
Comment (ginccc) [reply src/main/java/ai/labs/eddi/engine/memory/DeepCopyUtil.java]: valid point, good catch!

**Review by rolandpickl — APPROVED** [review]
